### DISCUSSION TITLE
Java - 1 - 성은, 진호

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#16] [질의 응답]
Java 질의응답

**Comment by seongeun42** [comment]
### ❓️자바 소스 파일의 컴파일 & 실행 과정에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

1. 소스 파일을 컴파일러가 Byte 코드로 이루어진 class 파일로 변환함
2. Class Loader에 의해 class 파일이 Loading, Linking, Initalizaion 세 단계를 거쳐 JVM의 메모리 영역에 동적으로 로딩됨
    1. Loading : class 파일을 JVM의 Method 영역에 저장
    2. Linking : verification(class 파일이 유효한지), preparation(class가 필요로 하는 메모리 할당하고 정의된 필드, 메소드, 인터페이스를 나타내는 데이터 구조 준비), resolution(심볼릭 레퍼런스를 실제 객체로 전환) 세 가지 기능을 수행
    3. Initalization : Static 필드들을 설정된 값으로 초기화함
3. Execution Engine이 Loading된 class 파일의 Byte 코드를 두 가지 방식으로 해석해 실행함
    1. 인터프리터 : Byte 코드 명령어를 하나씩 읽어서 해석하고 실행(한 줄의 실행은 빠르나 전체 실행은 느림)
    2. JIT컴파일러 : Byte 코드 전체를 컴파일해 Binary 코드를 변경 후 직접 실행함(빠름)

</blockquote>

<br>

**Reply by seongeun42** [reply]
#### ❓️ JVM의 메모리 구조에 대해 설명해주세요.
#### 💬 답변
<blockquote>

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/4d824417-d4ce-4ce6-a78b-b2dff555b4f7)

- **Method Area**
  - 모든 스레드가 공유함
  - 클래스의 멤버 변수의 이름, 데이터 타입, 접근 제어자 등의 각종 필드 정보와 메서드 정보, 데이터 Type 정보, static 변수, final class, Runtime Constant Pool 등 프로그램 실행할 때 필요한 데이터들이 저장되는 영역
  - Class Area나 Static Area로도 부름

- **Heap Area**
  - 모든 스레드가 공유함
  - new 키워드로 생성된 객체와 배열이 저장되는 영역
  - GC가 관리함

- **Stack Area**
  - 스레드 별로 생성됨
  - 함수 실행 시 지역 변수, 파라미터, 리턴 값 등 함수 내부에서 일시적으로 쓰이는 값들이 저장되는 영역
  - 함수가 호출될 때마다 그 함수를 위한 스택 프레임이 쌓이고, 종료되면 삭제됨

- **PC Register**
  - 스레드 별로 생성
  - 스레드가 생성될 때마다 생기는 영역으로, 현재 수행 중인 JVM 명령어 주소를 저장하고 있음
  - JVM 명령어 주소는 스레드가 실행 중인 부분을 기록하고 있음

- **Native Method stack**
  - Java 이외의 언어 혹은 기계어로 작성된 네이티브 코드를 실행할 때 사용되는 메모리 영역

</blockquote>

**Reply by sikutisa** [reply]
@seongeun42 @Jihn0118 
* 꼬리 질문에서 **Method Area가 GC에 의해 관리**된다라는 언급이 있는데, 근거 레퍼런스 있으면 추가해주셨으면 합니다.
* [오라클 자바 도큐먼트](https://docs.oracle.com/en/java/javase/20/gctuning/garbage-collector-implementation.html#GUID-23844E39-7499-400C-A579-032B68E53073)에서는 해당 내용을 못 찾았습니다(내용이 있는데 제가 못 찾은 걸 수도 있긴 합니다).

**Reply by seongeun42** [reply]
@sikutisa 
글 작성하다 잘못 들어간 거 같아요;; 수정했습니다.

**Comment by seongeun42** [comment]
### ❓️ Call By Value / Call By Reference에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>
Call By Value와 Call by Reference는 메서드를 호출할 때 파라미터를 전달하는 방법

- Call By Value
  - 호출자의 변수와 수신자의 변수의 파라미터는 복사 된 서로 다른 변수다.
  - 인자로 받은 값을 복사하여 처리
  - 복사하여 처리하기 때문에 안전하지만 메모리 사용량이 늘어남
  - 원래의 값이 보존

- Call By Reference
  - 참조를 직접 넘기기 때문에 호출자의 변수와 수신자의 파라미터는 동일한 변수다.
  - 인자로 받은 값의 주소를 참조하여 직접 값에 영향
  - 복사하지 않고 직접 참조를 하기에 빠름
  - 원래의 값에 영향
</blockquote>

<br>

**Reply by seongeun42** [reply]
#### ❓️ Java에서 사용되는 방식은 무엇인가요?
#### 💬 답변
<blockquote>

Java는 오로지 Call By Value를 사용함

</blockquote>

---

#### ❓️ Call By Value는 값을 복사해 넘기기 때문에 원래 값에 영향을 주지 않는데, Java에서 객체나 배열의 경우 파라미터로 넘겨도 원본 값에 영향을 줍니다. 이유가 무엇인가요?
#### 💬 답변
<blockquote>
객체나 배열의 경우 Reference Type인데, 참조 타입은 heap memory 영역에 생성된 객체의 주소 값을 참조하기 때문에 '참조 타입'이라고 불립니다.
따라서 파라미터로 넘겨준 것은 객체의 주소 값이고, 동일한 주소 값을 가지고 이 객체의 상태를 수정하면 당연히 동일한 주소를 참조하기 때문에 원본 값에 영향을 줍니다.

</blockquote>

**Comment by seongeun42** [comment]
### ❓️자바의 자료형에 대해 아는 대로 설명해주세요.

<br>

### 💬 답변
<blockquote>

크게 Primitive Type과 Reference Type으로 나뉨. Primitive는 실제 값을 저장하고, Reference는 값이 저장된 메모리 주소를 가리키는 값을 저장함. Primitive에는 boolean, char, byte, short, int, long, float, double 8가지가 있으며 범위가 작은 타입은 큰 타입으로 묵시적 형변환이 가능함. 반대의 경우 명시적 형변환이 가능하며 값의 손실이 일어날 수 있음. Reference는 크게 객체형과 배열형으로 나눌 수 있는데, 객체형은 클래스를 인스턴스화한 것을 가리키고, 배열형은 같은 타입을 값을 연속적인 공간에 여러 개를 모아둔 자료구조를 가리킴.

Primitive에는 boolean, char, byte, short, int, long, float, double 8가지가 있으며 범위가 작은 타입은 큰 타입으로 묵시적 형변환이 가능함. 반대의 경우 명시적 형변환이 가능하나 값의 손실이 일어날 수 있음. Primitive를 Reference형으로 바꾸는 Wrapper 객체가 있는데, 이들 간에는 오토박싱과 오토언박싱이 일어날 수 있음.

Reference는 크게 객체형과 배열형으로 나뉘는데, 객체형은 클래스를 실제 값을 가지는 객체 인스턴스화 했을 때 힙 영역에 생기는 실제 값을 가리키는 주소  참조값을 가짐. 배열형은 배열이 실제로 생긴 메모리 주소의 참조값을 가짐. 객체형의 경우 상속 관계에 있다면 형변환이 가능함.

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️Java Collection에 대해서 설명해주세요.

<br>

### 💬 답변
<blockquote>

Java Collection이란 데이터의 집합, 그룹을 의미하며 여러 자료구조를 구현해 놓은 클래스를 정의해둔 인터페이스를 제공함. 크게 `List`와 `Set`이 있는 `Collection`과 키와 값으로 이뤄진 `HashMap`을 가진 Map이 있음.

- Set : 순서를 유지하지 않고 중복을 허용하지 않는 집합 개념의 자료구조
   - SortedSet, HashSet, TreeSet, ...

- List : 순서가 있는 데이터의 집합으로, 중복을 허용함
   - ArrayList, LinkedList, ...

- Map : 키와 값 쌍으로 이루어진 데이터의 집합으로, 순서는 유지되지 않고, 키의 중복은 허용되지 않으나 값의 중복은 허용
   -  HashMap, TreeMap, ...

</blockquote>

<br>

**Reply by Jihn0118** [reply]
#### ❓️ 컬렉션에서 사용되는 제네릭에 대해서 설명해주세요.
#### 💬 답변
<blockquote>
제네릭(Generic)은 자바에서 데이터 타입을 파라미터화하여 클래스나 메서드를 일반화하는 기능을 말한다.

제네릭을 사용하면 컬렉션에 저장되는 요소들의 타입을 컴파일 시점에 체크할 수 있으므로, 타입 안정성을 보장할 수 있음
이를 통해 런타임 시점에서 발생할 수 있는 타입 관련 예외를 컴파일 시점에 미리 방지할 수 있음

</blockquote>

---

**Comment by seongeun42** [comment]
### ❓️JAVA에는 접근 제어자가 4가지 있습니다. 각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

<br>

### 💬 답변
<blockquote>
접근 제어자를 사용하면 변수나 메서드의 사용 권한을 설정할 수 있다.

JAVA의 접근 제어자 4가지
- private
- default
- protected
- public

**private**
private이 붙은 변수나 메서드는 해당 클래스 안에서만 접근이 가능하다.

**default**
접근 제어자를 별도로 설정하지 않는다면 변수나 메서드는 default 접근 제어자가 자동으로 설정되어 동일한 패키지 안에서만 접근이 가능하다.

**protected**
proteced가 붙은 변수나 메서드는 동일 패키지의 클래스 또는 해당 클래스를 상속 받은 클래스에서만 접근이 가능하다.

**public**
public 접근 제어자가 붙은 변수나 메서드는 어떤 클래스에서도 접근이 가능하다.

공개 범위: public > protected > default > private
### 의도치 않은 실수를 줄이고 정보 은닉의 목적으로 접근 제어자를 사용!

</blockquote>

<br>
